### PR TITLE
feat(fs): io_uring Fs implementation for high-throughput I/O

### LINKED ISSUE #77: feat: io_uring Fs implementation for high-throughput I/O
## Upstream References
- fjall-rs/lsm-tree#68 — Try io_uring inside SegmentWriter (compaction)
- TiKV io_uring experiment: https://openinx.github.io/ppt/io-uring.pdf

## Depends on
- #75 — Fs trait definition
- Does NOT depend on #76 (refactoring) — can develop in parallel

## Problem

lsm-tree uses synchronous `std::fs` I/O everywhere. On modern NVMe SSDs with deep I/O queues, this leaves performance on the table:
- Compaction writes: sequential, one-at-a-time I/O submissions
- Block reads: synchronous pread, no I/O batching
- fsync: blocking — stalls the calling thread

### IoUringFs

\`\`\`rust
/// io_uring-backed Fs implementation.
/// Linux only, feature-gated.
/// Sync interface wrapping async io_uring — dedicated I/O thread
/// with submission queue batching.
pub struct IoUringFs {
    ring: Arc<Mutex<IoUring>>,
    io_thread: JoinHandle<()>,
    submit_tx: crossbeam::channel::Sender<IoOp>,
    complete_rx: crossbeam::channel::Receiver<IoResult>,
}

impl Fs for IoUringFs {
    type File = IoUringFile;
    // ...
}
\`\`\`

### Key design decisions

1. **Sync interface** — `Fs` trait is sync. IoUringFs internally submits to ring and blocks on completion. Batching happens when multiple threads submit concurrently
2. **Dedicated I/O thread** — one thread owns the io_uring ring, polls completions, dispatches results
3. **Feature-gated** — `#[cfg(feature = "io-uring")]`, Linux only
4. **Runtime fallback** — if io_uring not available (old kernel), fall back to StdFs at startup with log warning
5. **Crate** — `io-uring` (low-level, no tokio dependency)

### What benefits most from io_uring

| Operation | Current | With io_uring | Expected gain |
|-----------|---------|---------------|---------------|
| Compaction write (MultiWriter) | write() + fsync per file | Batched SQE submissions | 2-3x throughput |
| Block read (table scan) | pread per block | readv / batched reads | 1.5-2x for large scans |
| fsync (journal, SST finalize) | Blocking per-file | io_uring FSYNC op | Reduced stall time |
| O_DIRECT reads | Not used | Native io_uring support | Bypass page cache for cold data |

### Scope

- [ ] `IoUringFs` struct with dedicated I/O thread
- [ ] `IoUringFile` with read/write/sync via ring submissions
- [ ] Runtime io_uring probe (kernel support detection)
- [ ] Fallback to StdFs if unsupported
- [ ] Benchmark: compaction throughput StdFs vs IoUringFs on NVMe
- [ ] Feature gate: `io-uring = ["dep:io-uring"]`

### Out of scope (future)
- Async Fs trait variant
- io_uring for network I/O (Raft, gRPC)
- Registered buffers / fixed files (optimization pass)

## Acceptance Criteria

- [ ] `IoUringFs` passes all lsm-tree tests on Linux
- [ ] Feature-gated: `cargo build` without `io-uring` feature works (no Linux dependency)
- [ ] Benchmark shows measurable improvement for compaction on NVMe
- [ ] Graceful fallback on kernels without io_uring support

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- `IoUringFs` implementing the object-safe `dyn Fs` trait with dedicated I/O thread and opportunistic SQE batching
- `IoUringFile` implementing `dyn FsFile` — routes read/write/fsync through the ring; cold-path ops (mkdir, stat, rename) delegate to `std::fs`
- `read_at` provides fill-or-EOF semantics with internal EINTR retry, matching the `FsFile` trait contract
- Runtime probe `is_io_uring_available()` for graceful fallback
- Feature-gated: `io-uring = ["dep:io-uring"]`, Linux-only target dependency
- 21 tests including concurrent `read_at` from 10 threads and edge-case coverage

## Design Decisions

- **No libc dependency for errno constants** — values like `EINTR (4)` and `EIO (5)` are inlined with comments, consistent with `StdFs` which uses raw FFI for `flock` without importing libc
- **Oversized buffers rejected with `InvalidInput`** — SQE length is `u32` but CQE result is `i32`, so buffers exceeding `i32::MAX` are rejected via `i32::try_from(buf.len())?.unsigned_abs()`. In practice LSM block I/O is 4-64 KB
- **Fatal ring error aborts the process** — if `submit_and_wait` fails (non-EINTR), previously submitted SQEs may still reference caller buffers. `std::process::abort()` is the only sound option
- **Ring thread panic aborts via `catch_unwind`** — if `event_loop` panics after submitting SQEs, those SQEs still reference caller buffers. `pending` map is wrapped in `ManuallyDrop` so SyncSenders survive stack unwinding, keeping callers blocked. `catch_unwind` + `abort` then kills the process before any buffer can be freed
- **Append mode uses `is_append` flag** — writes always query `file.metadata()?.len()` for the current EOF, ignoring the seek cursor. This matches O_APPEND semantics since io_uring uses explicit offsets
- **SQ full uses backpressure, not error** — when the submission queue is full, `enqueue` calls `submit_and_wait(1)` to drain a completion and retries the push. Since the Fs API is synchronous, callers are already blocking; backpressure is natural
- **`AtomicU64` for cursor** — could be plain `u64` (already `Sync`), kept for interior-mutability pattern consistency and potential future shared cursor access
- **Mutex on send_and_wait hot path** — guards `Option<SyncSender>` for clean shutdown. Lock held only for `send()` duration (~ns), negligible vs I/O latency (~µs) Submission channel is bounded to ring capacity (sync_channel) for natural backpressure
- **FxHash for pending map** — uses `crate::HashMap` (FxBuildHasher) for reduced hashing overhead on the I/O thread hot path
- **Seek positions may exceed `i64::MAX`** — matches `std::fs::File::seek` behavior; kernel rejects out-of-range offsets at the actual I/O syscall
- **Ring-thread error paths excluded from coverage** — `event_loop`, `enqueue`, and `Drop` contain error recovery (EINTR, SQ overflow, fatal ring failure, mutex poisoning) that requires kernel fault injection to exercise

## Test Plan

- [x] `cargo check` — clean build without `io-uring` feature (macOS/Windows)
- [x] `cargo test --lib` — all existing tests pass (no regressions)
- [x] `cargo test --lib --features io-uring` — 21 io_uring tests (requires Linux 5.6+)
- [x] Edge cases: empty buffers, seek overflow/underflow, sync_directory validation, Debug impl
- [ ] Benchmark: compaction throughput StdFs vs IoUringFs on NVMe

Closes #77

### CODE REVIEW

File: src/fs/io_uring_fs.rs [hunk 1]
```diff
@@ -0,0 +1,1221 @@
...
+/// Default number of `io_uring` submission queue entries.
+const DEFAULT_SQ_ENTRIES: u32 = 256;
+
+/// Probes whether `io_uring` is supported on the running kernel.
+///
+/// Creates a minimal ring with 2 entries and immediately drops it.
+/// Returns `false` on any failure (old kernel, seccomp restrictions, etc.).
+#[must_use]
+pub fn is_io_uring_available() -> bool {
+    IoUring::new(DEFAULT_SQ_ENTRIES).is_ok()
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/fs/io_uring_fs.rs]: @coderabbitai Stale — code was reverted to IoUring::new(2) and doc says '2-entry ring'. Both match.
